### PR TITLE
Quaternion based attitude control

### LINKED ISSUE #9077: 【mc_att_control】 Remove unnecessary defaults.
:smiley:  I don't think we need this extra variable.　https://github.com/PX4/Firmware/blob/master/src/modules/mc_att_control/mc_att_control_main.cpp#L95　
This variable `ATTITUDE_TC_DEFAULT`　is used to adjust the pid.　
Use it here:　https://github.com/PX4/Firmware/blob/master/src/modules/mc_att_control/mc_att_control_main.cpp#L534
I think we can use this parameter `MC_ROLL_TC`　to **directly** adjust the pid.　https://github.com/PX4/Firmware/blob/master/src/modules/mc_att_control/mc_att_control_params.c#L54
It is also a mistake to follow the current logic.　https://github.com/PX4/Firmware/blob/master/src/modules/mc_att_control/mc_att_control_params.c#L45
Should : **Increase** if the system is too twitchy, **Reduce** if the response is too slow and sluggish.

@MaEtUgR  @Stifael  Do I understand that correctly?　:rabbit2:

### PR DESCRIPTION
**Pull request by MaEtUgR**
**Introduction:** I had the plan to port and contribute the quaternion based attitude controller I wrote for my master thesis running on PX4 for too long now. I guess it was a big mistake to wait for https://github.com/PX4/Matrix/pull/34 to be merged which was pending for almost a year now. I was just too pigheaded for the incorrect quaternion convention because of a dramatic experience in the past. It's fixed and merged to PX4 with https://github.com/PX4/Firmware/pull/7908 now and hence I started my work.

**Documentation/Motivation:** Here's a really good paper documenting the underlying math behind this type of attitude control which is very similar but still far from equal to the well known rotation matrix based algorithm by Kumar et al..
https://www.research-collection.ethz.ch/bitstream/handle/20.500.11850/154099/eth-7387-01.pdf

I'm also unhesitant to share my master thesis for further documentation. It contains a hopefully even easier to read in depth text description of my point of view and includes my findings. Just let me know if and where it should be hosted.

**State:** ~~I started now by writing and testing the general quaternion attitude control law which with the current setpoint generation should generate the same result as running only the full attitude control of the paper and does not have any notion of reducing yaw authority. This was a lot simpler to implement as a first step and should already allow unrestricted flying. I checked stable multicopter flight in SITL. I will add the quaternion based roll pitch reduced attitude control and the mixing with the yaw now to enhance performance in extreme conditions.~~

- [x] Full attitude control (without lower yaw prioritisation)
- [x] Reduced attitude control (yaw neglected)
- [x] Mixing full and reduced
- [x] Logic to calculate yaw weight/priority from gains
- [x] Test and handle extreme corner cases
- [x] ~~Refactor attitude controller with matrix library (as suggested in review comments)~~
**EDIT:** I did only some refactoring and I think we should split it up here and make a separate PR for further such changes to keep it more focused and reviewable.

~~**Note:** Be **very careful** with tests on real hardware right now.~~

~~**Context:** This pr can soon be tested/compared with @bkueng's nice recent attitude enhancements:
https://github.com/PX4/Firmware/pull/7940
https://github.com/PX4/Firmware/pull/7997~~

**EDIT:** fixes #9077

### CODE REVIEW

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 1]
```diff
@@ -848,77 +850,26 @@ void
...
-	math::Vector<3> e_R = R.transposed() * (R_z % R_sp_z);
-
-	/* calculate angle error */
-	float e_R_z_sin = e_R.length();
-	float e_R_z_cos = R_z * R_sp_z;
-
-	/* calculate weight for yaw control */
-	float yaw_w = R_sp(2, 2) * R_sp(2, 2);
+	using namespace matrix;
+	float yaw_w = .4f;
```
Comment (dagar) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: Constant 0.4?

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 2]
```diff
@@ -848,77 +850,26 @@ void
...
-	/* axis and sin(angle) of desired rotation */
-	math::Vector<3> e_R = R.transposed() * (R_z % R_sp_z);
-
-	/* calculate angle error */
-	float e_R_z_sin = e_R.length();
-	float e_R_z_cos = R_z * R_sp_z;
-
-	/* calculate weight for yaw control */
-	float yaw_w = R_sp(2, 2) * R_sp(2, 2);
+	using namespace matrix;
```
Comment (dagar) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: You can do this at the top for all of mc_att_control.

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 3]
```diff
@@ -848,77 +850,26 @@ void
...
-		math::Vector<3> e_R_d = q_error(0) >= 0.0f ? q_error.imag()  * 2.0f : -q_error.imag() * 2.0f;
-
-		/* use fusion of Z axis based rotation and direct rotation */
-		float direct_w = e_R_z_cos * e_R_z_cos * yaw_w;
-		e_R = e_R * (1.0f - direct_w) + e_R_d * direct_w;
-	}
+	/* using sin(alpha/2) scaled rotation axis as attitude error (see quaternion definition by axis angle)
+	 * also taking care of the antipodal unit quaternion ambiguity */
+	Vector3f eq = 2.f * math::sign(qe(0)) * qe.imag();
+	math::Vector<3> e_R(eq.data());
```
Comment (dagar) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: Can you do this without mixing math libraries?

File: src/modules/mc_att_control/mc_att_control_main.cpp (hunk 1 above)
Comment (MaEtUgR) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: Yes, I can recommend you to read section 4.2 on page 16 of the paper I linked to about what this weight actually should represent and some comparison of resulting behaviours. I personally found 0.4 a good rule of thumb for a typical quadrotor setup in my past experience, that's why I put it there for the time being. I think this should be incorporated with the attitude yaw gain parameter in the end. Still to do.

It would be interesting where the current yaw weight calculation comes from and how it's justified from a control perspective. @bkueng told me already that it's done wrong and has his improvement suggestion here: https://github.com/PX4/Firmware/pull/7997/files#diff-81c4f50bdf837c4183b5d3cb81eb40eaR898

File: src/modules/mc_att_control/mc_att_control_main.cpp (hunk 3 above)
Comment (MaEtUgR) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: I got asked this before in reviews but in my opinion the matrix library is like the name suggests no general math library. I see the matrix library covering linear algebra and the math library more general math functions like sin, cos, sign, expo. That's why I also added all my useful mathematical helper functions here: https://github.com/PX4/Firmware/blob/13e64d00a8b8afa7cedd201ee85842de4feeff6a/src/lib/mathlib/math/Functions.hpp

Let me know if I got something wrong there.

Comment (dagar) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: I'm fine with the existence of the math library, but I don't like that we have multiple types of Vector. Vector3f is from Matrix and math::Vector<3> is from the PX4 math library.

Comment (MaEtUgR) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: I totally agree! I, like everyone else I talked to, favor the `matrix` library initiated by @jgoppert and that's why I contributed to it and look at `math::matrix` as deprecated.

Like in this pr I port every instance I come along to `matrix`. I also plan to get rid of the parameters:
https://github.com/PX4/Firmware/pull/8003/files#diff-81c4f50bdf837c4183b5d3cb81eb40eaR875
and outputs:
https://github.com/PX4/Firmware/pull/8003/files#diff-81c4f50bdf837c4183b5d3cb81eb40eaR872
in math::vector.

This is not to confuse with the call to `math::sign` which is a nonlinear scalar function that is defined in the `math` library because I felt it fits there best.

Comment (Stifael) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: I am bit confused as well. What is the reason for using math::Vector<3> instead of matrix?

**Comment by Stifael** [comment]
just a question about the implementation order. 
The way I understand full and reduced attitude control is to compute an attitude based on thrust vector and yaw. This means that it is not a control law, but rather a method to compute a desired quaternion.
In this function [here](https://github.com/PX4/Firmware/blob/8d4d3dcee58525552ac75af2ae9539ec0b4f03b6/src/modules/mc_att_control/mc_att_control_main.cpp#L844-L850), I expect a control law. In fact, the inputs to this function is a desired and a current quaternion, so it seems to be kind of confusing to read the line [here](https://github.com/PX4/Firmware/blob/8d4d3dcee58525552ac75af2ae9539ec0b4f03b6/src/modules/mc_att_control/mc_att_control_main.cpp#L866). I would expect such a description at the location where the desired attitude is actually computed.

**Comment by MaEtUgR** [comment]
Yes @Stifael you understood it completely right.

**Confusion:** I think the only problem with your confusion was my imprecise phrasing with the name "**full** attitude control" with which I refered to the current algorithm effect and not the exact formulas written in the paper under the corresponding section. Sorry for that I've seen these terms for too long now and start to use them vague. I edited the description now.

**Setpoint improvements:** We definitely should not only improve the controller but also the setpoint generation.
- Because now the controller needs to do back calculation to restore the reduced (roll/pitch, yaw) setpoint portions even though they were available before in form of the thrust vector. This inefficient back and forth is of course not done in the paper.
- The setpoint generation was suboptimal before like @bkueng already wrote and started to fix.

But these changes are bigger and could interfere also with fixed wing because of the required interface changes... We should tackle that step by step.

**Comment by Stifael** [comment]
confusion solved

File: src/modules/mc_att_control/mc_att_control_main.cpp (hunk 3 above)
Comment (MaEtUgR) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: It's there to stay compatible with the rest while not replacing all data types.

But I see the consensus is that I should not only replace the attitude control functionality but totally switch the entire module to the `matrix` library.

**Comment by MaEtUgR** [comment]
Rebased, it still flies super nicely in SITL. I already wrote soem scripts to do test calculations with the yaw adjustments and now I just need to port the correct implementation into the module.

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 4]
```diff
@@ -842,29 +839,35 @@ MulticopterAttitudeControl::control_attitude(float dt)
...
 	qd.normalize();
 
-	/* calculate reduced attitude which we would command if we about the vehicle's yaw */
+	/* calculate reduced desired attitude neglecting vehicle's yaw to prioritize roll and pitch */
 	Vector3f e_z = q.dcm_z();
 	Vector3f e_z_d = qd.dcm_z();
 	Quatf qd_red(e_z, e_z_d);
-	qd_red *= q;
+
+	if (Quatf(qd_red - Quatf(0, 1, 0, 0)).norm() > 1e-5f) {
```
Comment (bkueng) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: Use norm squared to avoid calling `sqrt`?

And maby I'd even add something like you added as commit message as code comment:
```
Corner case with arbitrary yaw but and 180 degree roll or pitch:
Reduced attitude control calculation always rotates around roll
because there's no right choice when neglecting yaw. In that small
corner case it's better to just use full attitude contol and hence
rotate around the most efficient roll/pitch combination.
```

Comment (MaEtUgR) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: 👍 Thanks for reviewing and the hints. You catched the one line which I'm still not pleased with. I think it's still better to have no condition at all than this condition. Everything works and is stable without this condition but if you make up a special case where the vehicle needs to rotate exactly (machine accuracy close) 180 degree around roll or pitch but no rotation around yaw the result is not the optimal control law. I will loose some sleep over it and hopefully find a good solution.

Comment (MaEtUgR) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: Found a better solution! What do you think?

**Comment by MaEtUgR** [comment]
All these changes were tested on a real vehicle today and it flew exactly as expected with a bit better yaw tracking like @bkueng already wrote.

**Comment by mhkabir** [comment]
Will be testing this soon on a racer. Great stuff! :)

**Comment by MaEtUgR** [comment]
Any reviews appreciated, in my opinion it's in a mergable state now.
(I can also share my low on comments matlab scripts if anyone wants to test out the math by himself...)

**Comment by MaEtUgR** [comment]
@bkueng Thanks a lot for your tests! If the vehicle attitude `q` or setpoint `qd` is [0, 0, 0, 0] then the output can be expected NaN because by definition an attitude quaternion has to have unit length. In this case there's no way for the controller to know the useful output. I would suggest that if the quaternion setpoint only is zero we instead take the identity quaternion and make the vehicle level and the rate controller treats NaN setpoint input as 0 for desperate failsafes.

The question is why are these quaternions ever zero. This general problem must have existed before and you just found it.

**Comment by MaEtUgR** [comment]
@bkueng I took your branch and debugged how NaNs get produced.
- On initialization `_v_att_sp` gets filled with zeros (that's ok)
- While not armed `mc_pos_control` does not publish any sttitude setpoint (not sure why that should be)
- The attitude control just uses the data in `_v_att_sp` if it was (ever) updated or not

The old controller just ignores the invalid setpoint and generates useless valid output and the new controller produces NaN in the same case. It's good that you found it, let's come up with a better general solution for this case. But it's not about the changes in this PR, they just reveal the problem that was always there.

**Comment by bkueng** [comment]
You're right. I think we should just initialize `q_d` with a valid identity quaterion.

But there's another NaN case, which is due to numerical inaccuracies. See https://github.com/bkueng/Firmware/commit/e47017a9f61fcb583ef7013a27695752e481ad8a, which contains a fix as well.

**Comment by MaEtUgR** [comment]
@bkueng Yes, you're right, I found multiple corner cases I'm just too slow in writing, the one with numerical values exceeding the domain of acos adn asin I also found as well thanks to your analysis branch and fixed exactly the same way. Now I wanted to initialize the `q` and `q_d` and found a strange case again depending on which values I take which I'm analyzing right now.

**Comment by bkueng** [comment]
I did some flight testing with the added `constrain`'s. It flies really well, the improved yaw tracking is also noticeable when flying sharp curves or fast small circles.
Here's a log (I switched a few times to the old controller): https://logs.px4.io/plot_app?log=2c353752-fec2-4f58-95d5-3bad7781c6d0

> SITL stops the entire attitude control loop reproducable within seconds/some minutes when it's just idle.

I tested that branch and let it run for around 10 minutes, but didn't have any issues. Maybe it's just too many printf's to the console?

**Comment by MaEtUgR** [comment]
@bkueng Thanks for all your super useful help! I just pushed fixes for the domain problem and initialization of the messages. I intentionally ommited checks for valid input quaternions on every loop iteration like https://github.com/MaEtUgR/Firmware/commit/a85c97cb891335afe822a40efb65c2fee66522d5#diff-81c4f50bdf837c4183b5d3cb81eb40eaR947 because that's too much overhead and possible warnings in my eyes. What do you think?

**Comment by bkueng** [comment]
> What do you think?

That's fine. Can you check CI, it fails due to style problems.

**Comment by MaEtUgR** [comment]
Astyle on my windows toolchain passed... I need to check what's going wrong.
Fixed.

**Comment by bkueng** [comment]
@PX4/testflights Can you test this please? On different multicopters in Manual, Position & Mission.

**Comment by santiago3dr** [comment]
Couple flights on the Pixracer (V4):
https://logs.px4.io/plot_app?log=3494ddf1-772c-4ca7-b956-372b18b19ae7
https://logs.px4.io/plot_app?log=9afc5d72-6a4a-41c3-bf4d-87b2ff5283ad

good flights; nice flight performance; no perceivable difference from master

**Comment by Avysuarez** [comment]
flight with pixhawk 1 V2. good flight, but i had some issues with baro.
https://review.px4.io/plot_app?log=edadf374-49d9-4c60-82ec-1339c86a2c74

flight with pixhawk 2.1 V3. good flight no isuuses.
https://review.px4.io/plot_app?log=b2a6a473-4d55-4292-889a-2ef2f4fb387a

flight with pixhawk 3 Pro V4pro. good flight, no issues.
https://review.px4.io/plot_app?log=1a90b8e3-d7df-4697-bbdc-c80e239f9671

**Comment by dannyfpv** [comment]
v3 pixhawk mini flight great.
https://review.px4.io/plot_app?log=d16407b8-ba40-4a90-8753-8692dfd08259

**Comment by MaEtUgR** [comment]
@bkueng I was expecting CPU load to decrease 😲 I actually concentrated to make all the quaternion calculations as efficient as possible... Probably this line https://github.com/PX4/Firmware/pull/8003/files#diff-81c4f50bdf837c4183b5d3cb81eb40eaR868 uses most cycles... but it's necessary to have linearisation for the yaw weight to have a correct implementation according to the paper.

**Comment by dagar** [comment]
> I was expecting CPU load to decrease ￼

I've found the actual math of the controllers is a tiny portion. https://github.com/PX4/Firmware/pull/8731#issuecomment-358851282
